whistl/classifier.py

```diff
@@ -3,106 +3,212 @@
 import argparse
 import logging
 import numpy as np
-import os
-import pickle
 import random
+import time
 
 import torch
+from torch.autograd import grad
 import torch.nn as nn
 import torch.optim as optim
 from torch.utils.data import DataLoader
+from tqdm import tqdm
 
 import dataset
 import model
 import util
 
 
-def count_correct(output, labels):
-    '''Calculate the number of correct predictions in the given batch'''
-    # This could be more efficient with a hard sigmoid or something,
-    # Performance impact should be negligible though
-    correct = 0
-    predictions = [1 if p > .5 else 0 for p in output]
-    for y, y_hat in zip(predictions, labels):
-        if y == y_hat:
-            correct += 1
-    return correct
+def compute_irm_penalty(loss, dummy_w):
+    ''' Calculate the invariance penalty for the classifier'''
+    dummy_grad = abs(grad(loss, dummy_w, create_graph=True)[0])
```

Comment:
Might be nice to have a short high-level explanation of what this is doing here. I know you linked to the paper, but it takes a bit of time/effort to read and understand.

Something like "Calculate the invariance penalty for the classifier. This is defined in the paper as the norm of the gradient of the loss function for an input of 1." (or something else if what I wrote doesn't make sense, I'm still not totally sure I understand it)

You could even write out the formula from the paper in some sort of pseudo-Latex (I do this sometimes: https://github.com/greenelab/netreg/blob/master/data_models.py#L405) if you think that makes things clearer.